src/mlpack/methods/ann/layer/write_memory_impl.hpp

```diff
@@ -0,0 +1,205 @@
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_WRITE_MEMORY_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_WRITE_MEMORY_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "write_memory.hpp"
+
+#include "../visitor/forward_visitor.hpp"
+#include "../visitor/backward_visitor.hpp"
+#include "../visitor/gradient_visitor.hpp"
+#include "../visitor/reset_cell_visitor.hpp"
+#include "../visitor/forward_with_memory_visitor.hpp"
+#include "../visitor/backward_with_memory_visitor.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template <typename InputDataType, typename OutputDataType>
+WriteMemory<InputDataType, OutputDataType>::WriteMemory(
+  const size_t inSize,
+  const size_t numMem,
+  const size_t memSize,
+  const size_t shiftSize) :
+    inSize(inSize),
+    numMem(numMem),
+    memSize(memSize),
+    shiftSize(shiftSize)
+{
+  inputToLinear = new Linear<>(inSize, 2 * memSize);
+
+  addGate = new TanHLayer<>();
+  eraseGate = new SigmoidLayer<>();
+
+  writeHead = new MemoryHead<>(inSize, numMem, memSize, shiftSize);
+
+  network.push_back(inputToLinear);
+  network.push_back(addGate);
+  network.push_back(eraseGate);
+  network.push_back(writeHead);
+
+  dWriteHead.set_size(numMem, 1);
+  prevError.set_size(2 * memSize, 1);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void WriteMemory<InputDataType, OutputDataType>::ForwardWithMemory(
+    arma::Mat<eT>&& input, const arma::Mat<eT>&& memory, arma::Mat<eT>&& output)
+{
+  // Pass through linear layer.
+  boost::apply_visitor(ForwardVisitor(std::move(input), std::move(
+      boost::apply_visitor(outputParameterVisitor, inputToLinear))),
+      inputToLinear);
+
+  arma::mat& lOutput = boost::apply_visitor(outputParameterVisitor,
+      inputToLinear);
+
+  // Generate AddVec
+  boost::apply_visitor(ForwardVisitor(std::move(lOutput.submat(0, 0,
+      memSize - 1, 0)), std::move(boost::apply_visitor(outputParameterVisitor,
+      addGate))), addGate);
+  const arma::mat& addVec = boost::apply_visitor(outputParameterVisitor,
+      addGate);
+
+  // Generate EraseVec
+  boost::apply_visitor(ForwardVisitor(std::move(lOutput.submat(memSize, 0,
+      2 * memSize - 1, 0)), std::move(boost::apply_visitor(
+      outputParameterVisitor, eraseGate))), eraseGate);
+  const arma::mat& eraseVec = boost::apply_visitor(outputParameterVisitor,
+      eraseGate);
+
+  // Generate write weights
+  boost::apply_visitor(ForwardWithMemoryVisitor(std::move(input),
+      std::move(memory),
+      std::move(boost::apply_visitor(outputParameterVisitor, writeHead))),
+      writeHead);
+  const arma::mat& writeWeights = boost::apply_visitor(outputParameterVisitor,
+      writeHead);
+
+  output = memory;
+
+  auto addVecIt = addVec.begin();
+  auto eraseVecIt = eraseVec.begin();
+
+  output.each_col([&](arma::vec& v)
```

Comment:
I'm wondering if you remove the first ``&`` if that would solve the windows problem.